### PR TITLE
⚡️ Speed up function `gfo2hyper` by 15%

### PR DESCRIPTION
**Pull request by misrasaurabh1**
### 📄 15% (0.15x) speedup for ***`gfo2hyper` in `src/hyperactive/optimizers/constraint.py`***

⏱️ Runtime :   **`243 microseconds`**  **→** **`211 microseconds`** (best of `148` runs)
<details>
<summary> 📝 Explanation and details</summary>

To optimize the given program for better performance, we will focus on refining the loop inside the `gfo2hyper` function. One potential optimization is to avoid calling `search_space.keys()` repeatedly and to directly iterate over `search_space.items()`. This can help save time, especially for larger dictionaries.

### Changes Made.
1. Changed `for _, key in enumerate(search_space.keys()):` to `for key, values in search_space.items():` to directly access keys and associated values.

### Reason for Changes.
- Accessing both key and value directly in the same loop reduces time complexity and avoids extra dictionary lookups, leading to performance improvements.

</details>

✅ **Correctness verification report:**

| Test                        | Status            |
| --------------------------- | ----------------- |
| ⚙️ Existing Unit Tests | 🔘 **None Found** |
| 🌀 Generated Regression Tests | ✅ **12 Passed** |
| ⏪ Replay Tests | 🔘 **None Found** |
| 🔎 Concolic Coverage Tests | 🔘 **None Found** |
|📊 Tests Coverage       | 100.0% |
<details>
<summary>🌀 Generated Regression Tests Details</summary>

```python
import pytest  # used for our unit tests
from hyperactive.optimizers.constraint import gfo2hyper

# unit tests

# Basic Functionality Tests
def test_simple_case():
    search_space = {'a': [1, 2, 3], 'b': [4, 5, 6]}
    para = {'a': '1', 'b': '2'}
    expected = {'a': 2, 'b': 6}
    codeflash_output = gfo2hyper(search_space, para)

def test_single_key():
    search_space = {'a': [1, 2, 3]}
    para = {'a': '0'}
    expected = {'a': 1}
    codeflash_output = gfo2hyper(search_space, para)

# Edge Case Tests


def test_non_existent_key_in_search_space():
    search_space = {'a': [1, 2, 3]}
    para = {'b': '0'}
    with pytest.raises(KeyError):
        gfo2hyper(search_space, para)

# Invalid Indices Tests
def test_non_integer_index():
    search_space = {'a': [1, 2, 3]}
    para = {'a': 'x'}
    with pytest.raises(ValueError):
        gfo2hyper(search_space, para)

def test_out_of_bound_index():
    search_space = {'a': [1, 2, 3]}
    para = {'a': '5'}
    with pytest.raises(IndexError):
        gfo2hyper(search_space, para)


def test_multiple_keys_with_mixed_indices():
    search_space = {'a': [1, 2, 3], 'b': [4, 5, 6], 'c': [7, 8, 9]}
    para = {'a': '2', 'b': '1', 'c': '0'}
    expected = {'a': 3, 'b': 5, 'c': 7}
    codeflash_output = gfo2hyper(search_space, para)

def test_nested_lists():
    search_space = {'a': [[1, 2], [3, 4]], 'b': [[5, 6], [7, 8]]}
    para = {'a': '1', 'b': '0'}
    expected = {'a': [3, 4], 'b': [5, 6]}
    codeflash_output = gfo2hyper(search_space, para)

# Large Scale Test Cases
def test_large_search_space_and_para():
    search_space = {'key' + str(i): list(range(100)) for i in range(1000)}
    para = {'key' + str(i): str(i % 100) for i in range(1000)}
    expected = {'key' + str(i): i % 100 for i in range(1000)}
    codeflash_output = gfo2hyper(search_space, para)

# Performance and Scalability Tests

def test_high_number_of_keys():
    search_space = {'key' + str(i): [i] for i in range(1000)}
    para = {'key' + str(i): '0' for i in range(1000)}
    expected = {'key' + str(i): i for i in range(1000)}
    codeflash_output = gfo2hyper(search_space, para)

# Special Cases Tests


import pytest  # used for our unit tests
from hyperactive.optimizers.constraint import gfo2hyper

# unit tests

# Basic Functionality
```

</details>

To edit these changes `git checkout codeflash/optimize-gfo2hyper-m8ft3mp4` and push.

[Codeflash](https://codeflash.ai)

### CODE REVIEW

**Comment by misrasaurabh1** [comment]
@23pointsNorth what code formatting settings does this repo use? I could not find one so just defaulted to `black` but that sometimes causes some differences.

**Comment by SimonBlanke** [comment]
> @23pointsNorth what code formatting settings does this repo use? I could not find one so just defaulted to `black` but that sometimes causes some differences.

I always autoformat my files to 'black' on save. But there might be some differences in older files.

File: src/hyperactive/optimizers/constraint.py [hunk 1]
```diff
@@ -5,9 +5,9 @@
 
 def gfo2hyper(search_space, para):
     values_dict = {}
-    for _, key in enumerate(search_space.keys()):
+    for key, values in search_space.items():
```
Comment (SimonBlanke) [inline src/hyperactive/optimizers/constraint.py:8]: This requires the `items`-method in [this dict base-class](https://github.com/SimonBlanke/Hyperactive/blob/master/src/hyperactive/search_space.py#L8) to work.

**Review by SimonBlanke — CHANGES_REQUESTED** [review]

**Comment by fkiraly** [comment]
@misrasaurabh1, how did you measure the speed-up if the function actually breaks?

Is this PR completely AI generated and you did not even check whether it runs, or how should I understand this?

**Comment by misrasaurabh1** [comment]
Hi @fkiraly , the speed up is measured by codeflash using the generated test cases that are attached in the PR description. Looking at the code it was hard to know what input type it was going to get so it made an assumption that it would get a dictionary, and did the correctness and performance checks. 
I manually reviewed the optimization and since it looked good, i opened the PR here. But its true that the benchmark hasn't been run again with the modified code, but i believe that the performance gains should still hold

**Review by SimonBlanke — APPROVED** [review]
